### ISSUE TITLE
Spring hangs a lot

### ISSUE THREAD

**Issue by mockdeep** [rails/spring#396]
We've been having a lot of problems with spring hanging on our systems. I started outputting to a log and here's what I see:

```
[2015-03-19 14:27:01 -0700] [21417] [application:test] initialized -> running
/home/fletch/.rvm/gems/ruby-2.1.5/gems/spring-1.3.3/lib/spring/application.rb:286:in `read': stream closed (IOError)
    from /home/fletch/.rvm/gems/ruby-2.1.5/gems/spring-1.3.3/lib/spring/application.rb:286:in `block (2 levels) in with_pty'
[2015-03-19 14:27:01 -0700] [12838] [application_manager:test] child 21417 shutdown
[2015-03-19 14:27:01 -0700] [21531] [application:test] preloading app
[2015-03-19 14:27:06 -0700] [21531] [application:test] initialized -> running
/home/fletch/.rvm/gems/ruby-2.1.5/gems/spring-1.3.3/lib/spring/application.rb:286:in `read': stream closed (IOError)
    from /home/fletch/.rvm/gems/ruby-2.1.5/gems/spring-1.3.3/lib/spring/application.rb:286:in `block (2 levels) in with_pty'
```

I haven't got a strong idea of what is causing it just yet, though it seems to come up if we change git branches and the like, so maybe some sort of file changes are causing it to just lock up. We have to run `spring stop` and start over.

**Comment by ulfmagnetics** [comment]
+1 also seeing a lot of hangs on OS X Yosemite.

**Comment by mockdeep** [comment]
I'm on Linux Mint, though other users at our company are seeing it on Mac as well. Something as simple as changing branches can cause it:

```
[2015-03-26 12:20:36 -0700] [10899] [application:development] running -> watcher_stale
[2015-03-26 12:20:36 -0700] [10899] [application:development] watcher_stale -> exiting
[2015-03-26 12:20:36 -0700] [10896] [application_manager:development] child 10899 shutdown
[2015-03-26 12:20:36 -0700] [11147] [application:development] preloading app
[2015-03-26 12:20:42 -0700] [11147] [application:development] initialized -> running
/home/fletch/.rvm/gems/ruby-2.1.5/gems/spring-1.3.3/lib/spring/application.rb:286:in `read': stream closed (IOError)
	from /home/fletch/.rvm/gems/ruby-2.1.5/gems/spring-1.3.3/lib/spring/application.rb:286:in `block (2 levels) in with_pty'
[2015-03-26 12:20:53 -0700] [10896] [server] accepted client
[2015-03-26 12:20:53 -0700] [10896] [server] running command rails_console
[2015-03-26 12:20:53 -0700] [11180] [client] sending command
```

**Comment by mockdeep** [comment]
Ctrl+C won't get me out of it, either. I have to kill the process from elsewhere.

**Comment by ulfmagnetics** [comment]
Yes, this is exactly my issue. When I log to a file by setting SPRING_LOG, then change branches, I see the above sequence of messages appear (```running -> watcher_stale```, ```watcher_stale -> exiting```, ```preloading app```, etc.) After that, a call to ```bin/rspec``` hard-hangs and has to be killed from another shell.

**Comment by mscoutermarsh** [comment]
This is happening to me as well. For now I've added a post-checkout hook to call `spring stop`.

**Comment by whatasunnyday** [comment]
I am also experiencing this problem. I have to kill Spring processes that are consuming a lot of CPU even though `bin/spring stop` says there are no spring processes running.

**Comment by bborn** [comment]
+1 Spring status says Spring is not running (because I stopped it), but ps shows Spring processes still going.

**Comment by terracatta** [comment]
> though it seems to come up if we change git branches and the like, so maybe some sort of file changes are causing it to just lock up

+1

Can confirm switching branches in our app (which has "a lot" of stuff in `/lib`) causing spring to hang even on rake commands. I have to run `spring stop` in another terminal tab to kill the process so I can proceed. Also considering a post-checkout hook to call `spring stop`

**Comment by jafrog** [comment]
After a lot of tinkering I was able to figure out why it happens, at least in my case. 

First, Spring hangs when checking out a branch that has changes in one of the monitored files: `Gemfile` or config files. In this case Spring reloads an application on background. If your application **prints any warnings** on initialise the process will wait to access `STDOUT` which never happens as Spring's `application` process runs on background. In my case the solution was to disable warnings with `$VERBOSE=nil` on top of the `config/application.rb` because the app wasn't trying to print anything but warnings. But of course it won't do if you want to keep seeing warnings.

**Comment by jonleighton** [comment]
@jafrog can you point to the line of code where spring hangs?

**Comment by jafrog** [comment]
@jonleighton For me it was [here](https://github.com/rails/spring/blob/master/lib/spring/application.rb#L82), right when Rails app is loaded

**Comment by jafrog** [comment]
Looks like the bug was related to `STDERR` buffer overflow with warnings and was fixed in 1.3.0 ([acceptance test](https://github.com/rails/spring/blob/master/lib/spring/test/acceptance_test.rb#L127-L135)) I was able to reproduce it with 1.2.0 but 1.3.6 works fine.

**Comment by mockdeep** [comment]
Hmm, I'm still seeing some hangs. I just changed branches and ran `spring rake db:migrate` and after a few minutes nothing had happened. Running version `1.3.6`. Here are the logs just before the most recent lock up. The stack trace is when I ran `spring stop` from another terminal:

```
[2015-05-27 12:02:20 -0700] [20277] [server] accepted client
[2015-05-27 12:02:20 -0700] [20277] [server] running command rake
[2015-05-27 12:02:20 -0700] [20708] [client] sending command
[2015-05-27 12:02:20 -0700] [20281] [application:development] got client
[2015-05-27 12:02:20 -0700] [20281] [application:development] forked 20712
[2015-05-27 12:02:20 -0700] [20277] [application_manager:development] got worker pid 20712
[2015-05-27 12:02:20 -0700] [20708] [client] got pid: 20712
[2015-05-27 12:02:36 -0700] [20281] [application:development] 20712 exited with 0
[2015-05-27 12:02:36 -0700] [20708] [client] got exit status 0
[2015-05-27 12:29:23 -0700] [20281] [application:development] running -> watcher_stale
[2015-05-27 12:29:23 -0700] [20281] [application:development] watcher_stale -> exiting
[2015-05-27 12:29:23 -0700] [20277] [application_manager:development] child 20281 shutdown
[2015-05-27 12:29:23 -0700] [22234] [application:development] preloading app
[2015-05-27 12:29:29 -0700] [22234] [application:development] initialized -> running
/home/fletch/.rvm/gems/ruby-2.1.6/gems/spring-1.3.6/lib/spring/application.rb:286:in `read': stream closed (IOError)
	from /home/fletch/.rvm/gems/ruby-2.1.6/gems/spring-1.3.6/lib/spring/application.rb:286:in `block (2 levels) in with_pty'
[2015-05-27 12:29:38 -0700] [20277] [server] accepted client
[2015-05-27 12:29:38 -0700] [20277] [server] running command rake
[2015-05-27 12:29:38 -0700] [22310] [client] sending command
[2015-05-27 12:32:26 -0700] [20277] [server] shutting down
[2015-05-27 12:32:26 -0700] [20277] [application_manager:development] stopping
/home/fletch/.rvm/gems/ruby-2.1.6/gems/spring-1.3.6/lib/spring/application_manager.rb:22:in `unlock': Attempt to unlock a mutex which is locked by another thread (ThreadError)
	from /home/fletch/.rvm/gems/ruby-2.1.6/gems/spring-1.3.6/lib/spring/application_manager.rb:22:in `ensure in synchronize'
	from /home/fletch/.rvm/gems/ruby-2.1.6/gems/spring-1.3.6/lib/spring/application_manager.rb:22:in `synchronize'
	from /home/fletch/.rvm/gems/ruby-2.1.6/gems/spring-1.3.6/lib/spring/application_manager.rb:39:in `with_child'
	from /home/fletch/.rvm/gems/ruby-2.1.6/gems/spring-1.3.6/lib/spring/application_manager.rb:60:in `run'
	from /home/fletch/.rvm/gems/ruby-2.1.6/gems/spring-1.3.6/lib/spring/server.rb:59:in `serve'
	from /home/fletch/.rvm/gems/ruby-2.1.6/gems/spring-1.3.6/lib/spring/server.rb:44:in `block in start_server'
	from /home/fletch/.rvm/gems/ruby-2.1.6/gems/spring-1.3.6/lib/spring/server.rb:44:in `loop'
	from /home/fletch/.rvm/gems/ruby-2.1.6/gems/spring-1.3.6/lib/spring/server.rb:44:in `start_server'
	from /home/fletch/.rvm/gems/ruby-2.1.6/gems/spring-1.3.6/lib/spring/server.rb:38:in `boot'
	from /home/fletch/.rvm/gems/ruby-2.1.6/gems/spring-1.3.6/lib/spring/server.rb:14:in `boot'
	from -e:1:in `<main>'
[2015-05-27 12:32:26 -0700] [22310] [client] got no pid
```

**Comment by sgonyea** [comment]
Even on a brand new Rails project, I still see orphaned Spring 'app' processes hanging around, sipping around 6% CPU.  I have to `kill -9` them.  This is without doing any sort of git branching.

In an older (Rails 3) project, I see the same thing but with git branch I get orphaned 'server' processes too.

**Comment by grosser** [comment]
Also saw that but was never able to reproduce

**Comment by fl00r** [comment]
One of two times it hangs on El Capitan

**Comment by mockdeep** [comment]
We're still seeing spring hanging a lot both on Linux and Macos.

**Comment by aaronchi** [comment]
I did a bunch of debugging today on an app where spring was nearly unusable and it turns out that removing the `dnsruby` gem from the Gemfile solved the problem. Not sure what the cause of the issue is but hope this helps someone.

**Comment by inspire22** [comment]
Happening to me on mac as well.  I run from guard and sometimes spring will hang - even after restarting guard I won't get any results back until I issue "spring stop".  Might be loosely correlated with when I cancel tests that are running using ctrl+C (which I'm not sure how guard passes through to spring).  Guard also has a bug where ctrl+c will stop the tests running, but also stop guard from watching any files, but that's not related to you...

**Comment by pwim** [comment]
I can confirm that the `dnsruby` gem and Spring don't play well together. Removing it also prevents this hanging.

**Comment by arjun810** [comment]
@aaronchi we're still seeing a bunch of hangs, but aren't using dnsruby. Would you be able to share how you were debugging this? Thanks!

**Comment by ayufan** [comment]
You can try my fix: https://github.com/rails/spring/pull/484. It should help with hanging `rails console` :)

**Comment by markiz** [comment]
For us spring freezes during the second preload if an app is whining about already existing constant. I assume this is somehow connected with writing to an STDERR with a stale socket.

**Comment by markiz** [comment]
Okay, I've played with it a little and here's what I found.
Spring freezes during eager preload. That happens when PTY that we provided gets written more than X bytes without being read from the other end. Normally, we would read that using the other thread that we spawn. However, when ruby issues warnings like "previous definition of XXX was here", it uses blocking writes. Blocking write starts writing and PTY gets clogged. We can't do anything until we read something. But our reader is in the parallel thread. Which we never get to because we're currently trying to write something to stderr.

I don't know enough about why pty was introduced in the first place, however, as a workaround, I'm going to redirect `STDERR` to `/dev/null` in situations when we can't see it (e.g. preloading in the background).

Hope that made any sense.

**Comment by markiz** [comment]
https://github.com/markiz/spring/commit/5f3ab734fc45d541aaaccb05b59cd95aa49fe8ef this is what fixes freezes for us, sorry that I can't wrap it in a nice PR with tests and stuff for now.

**Comment by ledbettj** [comment]
@markiz In my limited testing that also solves a fairly reproducible hang for me as well.

**Comment by jmacdonald** [comment]
@markiz that solved it for me as well (Spring would also spike CPU usage to 100%); thanks!

**Comment by sbleon** [comment]
@markiz that's several 👍 s on your change! I think you should submit a PR for your branch. This may be the kind of thing that's too hard to test, so if that's holding things up don't worry about the tests.

**Comment by stwr667** [comment]
👍  @sbleon and @markiz . We would love this to be fixed in the main branch.

**Comment by sbleon** [comment]
I'm using the most recent release and no longer seeing hangs. Someone who is seeing hangs without the patch should submit a PR with the change.

**Comment by muriloime** [comment]
I am having the same issue (I do have dnsruby gem in my project), either with @markiz solution or 1.7.2 version. Anyone else seeing the same?

**Comment by JochenFromm** [comment]
I also noticed that a new [Rails 5.0 applications hangs](http://stackoverflow.com/questions/39344542/rails-app-using-devise-hangs-completed-401-unauthorized-in-x-ms/39876625) and freezes and could not solve the problem for weeks until I commented out the `spring` gem. It only appeared in development environment where the `spring` gem was included. The normal logs say nothing, the application just hangs, which is extremely nasty and annoying.

**Comment by misterbyrne** [comment]
This is happening all the time for a private project that I work on. Here's how I can recreate it at will:

Terminal 1 
`spring stop; spring server;`

Terminal 2
`rails runner 'puts 1 + 1'`
`touch config/application.rb`

You will see a `socket closed` exception logged to terminal 1 and the spring server will hang. Any subsequent spring commands will hang indefinitely.

This isn't the case for all projects though 😞

**Comment by misterbyrne** [comment]
I've recreated the problem with a test case in this PR https://github.com/rails/spring/pull/497

**Comment by aried3r** [comment]
We still see a lot of slowdowns on the initial boot after upgrading to Rails 5 and spring 2.

rails (5.0.1)
spring (2.0.1)
spring-commands-rspec (1.0.4)
spring-commands-rubocop (0.2.0)
spring-watcher-listen (2.0.1)

`DISABLE_SPRING=1 rails console` takes about 8 seconds.
`rails console` takes 75 seconds (1:15m). Subsequent calls are fast, since spring is now running.

Is there a way to have spring verbosely output what it's doing and find where it's hanging?

**Comment by rubys** [comment]
I don't know if this is the same bug, and while it is an odd scenario, it is 100% reproducible with Rails 5.1, Spring 2.0.2, and Mac OS/X High Sierra:

```
rails new project; cd project; bin/rails generate scaffold Product name; cd ..; rm -rf project; rails new project; cd project; bin/rails generate scaffold Product name
```

Generally, control-C isn't enough to stop the process, it requires killing.  Once killed, retrying the generate scaffold command once again hangs, *unless* a `spring stop` is done before then (yes, in the newly recreated project directory).

Hope this helps.

**Comment by guilpejon** [comment]
I'm having the exact same problem with Rails 5.0.2 and Spring 2.0.2 on my mac. When I run `rails console` the process hangs, I can't stop it with CMD+C, and I have to kill the process with the kill command. Sometimes the kill command is not enough and I have to restart my mac or comment out the spring gem.

**Comment by alexfulsome** [comment]
Spring 2.0.2, Rails 5.1.4. Happens so much that I have an alias for `spring stop`

**Comment by k00ka** [comment]
Rails 5.1.5, Spring 2.0.2 hangs more often than not when running 'rails c'.

**Comment by fredngo** [comment]
Seeing spring hang a lot with a new Rails 5.2.1 project on macOS 10.13.5 High Sierra

**Comment by amol-edmodo** [comment]
hangs a lot .... on ruby 2.5, rails 5.1.3
tried both spring 1.7.2 and 2.0.2

**Comment by amolpujari** [comment]
hangs a lot .... on ruby 2.5, rails 5.1.3
tried both spring 1.7.2 and 2.0.2

**Comment by JoelNichols** [comment]
Can confirm I'm having this issue on Rails 5.2. Getting a bit tiring having to kill the Rails process every so often.

Thought I was going mad until I found this issue.

**Comment by jgomo3** [comment]
2020. Still a problem. I have to stop it in order to be able to work with the Rails console.

**Comment by stwr667** [comment]
Try Zeus instead - https://github.com/burke/zeus
We've been using it for a more than a year in our workplace. It's not perfect, but more stable than Spring at least.

**Comment by MichaelSp** [comment]
`rails test` hangs but

`bin/rails test` works fine.

**Comment by mvz** [comment]
I had spring hang regularly, sometimes every five minutes. I've manually applied the patch in https://github.com/markiz/spring/commit/5f3ab734fc45d541aaaccb05b59cd95aa49fe8ef and haven't' seen any hangs since. As per https://github.com/rails/spring/issues/396#issuecomment-248906237 I'm going to submit that patch as a PR. Conveniently, it's also on a branch.

**Comment by mvz** [comment]
I'm still seeing this problem with the new spring release 2.1.1.

**Comment by growthcode** [comment]
My issue came from a file inside of bin `bin/spring`.
Everything worked once it was removed.

I don't know how it got there. I checked my command history and I never ran `spring binstub` which is how the file can be autogenerated.

**Comment by owen-le-goldenowl** [comment]
Jan 2021, it's seems the issue still exists. When I start rails c, I see a ruby proccess that comsumes 100% cpu. The proccess is not killed even I exit rails c
<img width="981" alt="Screen Shot 2021-01-15 at 10 02 51" src="https://user-images.githubusercontent.com/54341400/104676005-2581e200-5719-11eb-8f5a-c8462ff84a5d.png">
.

**Comment by andypple** [comment]
I added 2 gems into my Gemfile and it works
```
  gem 'listen', '~> 3.4.1'
  gem 'spring-watcher-listen'
```
Make sure using listen 3.4.1 or above.

**Comment by jaydorsey** [comment]
Not to pile on, but this is definitely reproducible with a minimal rails application.

```
rails new --skip-webpack-install --api test_spring && cd test_spring   
echo "ENV.fetch('FOOBAR')" > config/initializers/foo.rb  
bin/rails c # This hangs, prevents CTRL-C, CTRL-\, etc. from working
DISABLE_SPRING=1 bin/rails c # this works fine
```

Tested this with Ruby 3.0.0 but have seen the same behavior for a long time with other ruby versions.

I tried with the #693 fork changes and it didn't seem to fix this particular issue.

I did find something just now that might be interesting/relevant: this error only happens when I'm inside tmux. I wonder how many other people in this thread are using tmux?

**Comment by mcfoton** [comment]
Maybe this will help someone: while in the beginning I thought that `spring` is the issue, in my specific case ruby process eating 100% CPU was solved by explicitly updating `listen` gem version used in `spring-watcher-listen` gem that we had in the project.

**Comment by owen-le-goldenowl** [comment]
> Maybe this will help someone: while in the beginning I thought that `spring` is the issue, in my specific case ruby process eating 100% CPU was solved by explicitly updating `listen` gem version used in `spring-watcher-listen` gem that we had in the project.

Worked. Thank you 👍

**Comment by tbhockey** [comment]
> Not to pile on, but this is definitely reproducible with a minimal rails application.
> 
> ```
> rails new --skip-webpack-install --api test_spring && cd test_spring   
> echo "ENV.fetch('FOOBAR')" > config/initializers/foo.rb  
> bin/rails c # This hangs, prevents CTRL-C, CTRL-\, etc. from working
> DISABLE_SPRING=1 bin/rails c # this works fine
> ```
> 
> Tested this with Ruby 3.0.0 but have seen the same behavior for a long time with other ruby versions.
> 
> I tried with the #693 fork changes and it didn't seem to fix this particular issue.
> 
> I did find something just now that might be interesting/relevant: this error only happens when I'm inside tmux. I wonder how many other people in this thread are using tmux?
> 
> I typically have no other problems with tmux and tmux itself isn't hung up; even when it hangs I can create a new buffer & then issue `pkill -9 ruby` to kill the hanging process
> 
> Edit: I've also tried updating the listen gem and have seen mixed results

Interesting. I too am on tmux and experiencing this after recently bumping to Ruby 2.6.8.

**Comment by jaydorsey** [comment]
I tested [this](https://github.com/rails/spring/pull/623) fix (saw that github shows "May be fixed by 623")

Inside of tmux I still see the same issue. It still may fix _this_ issue (not sure if my issue is the same as reported by OP)

I tried my snippet posted earlier with the latest versions of things:

- ruby 3.0.2
- rails 6.1.4
- listen 3.6.0
- spring 2.1.1
- bootsnap 1.7.7

Not sure which versions I used originally, but since now I'm also seeing it hang _outside_ of tmux as well I wish I'd documented the versions so I could compare.

I'm happy to pair on this if anyone wants to, but I'm not even sure where to start with this. My tmux config is available online (as is most of my setup/shell)

**Comment by jgomo3** [comment]
Kind of related, and kind of good news: Spring is not default anymore https://github.com/rails/rails/pull/42997

**Comment by SaimonL** [comment]
I am having the same issue where Ctrl+c won't work and I have to kill it from task manager by force.

- macOS Big Sur (11.5.2)
- Rails 6.1.4.1 AND Rails 5.2.6
- Ruby 2.7.4p191 (2021-07-07 revision a21a3b7d23) [x86_64-darwin20] via rbenv

The worst part is that it is random so you don't know when it will happen and once it does from that point on its 100% of the time until you do one of the followings below:

1. If I disable spring then it loads fast and never gets stuck.
2. If I don't disable spring then I have to enable "eager_load"

**Comment by jussijarvio** [comment]
Experienced similar hanging with every "rails"-command on Rails 6.1.5 and after trying all kind of sorcery and witchcraft I ended up regenerating binstubs with "**rake app:update:bin**". This solved all "rails"-command issues for me.

**Comment by fredngo** [comment]
Wow, 7 years later, finally fixed!

**Comment by jasonfb** [comment]
I noticed recently that sometimes Spring hangs if you try to run a test (Minitest) with the wrong filename, for example, a test file that doesn't exist.
